src/validator.business.ts

```diff
@@ -0,0 +1,56 @@
+import { NIFType, CustomArgs } from './validator.model';
+
+const getExtra = (char: string) => (char != 'Z' ? (char != 'Y' ? 0 : 1) : 2);
+
+const validateDNIControlLetter = (letter: string, num: number) =>
+  letter === 'TRWAGMYFPDXBNJZSQVHLCKE'.charAt(num % 23);
+
+const checkDNI = (nif: string): NIFType => {
+  const extra = getExtra(nif.charAt(0));
+  const num = Number(`${extra}${nif.match(/\d+/)}`);
+
+  return validateDNIControlLetter(nif.charAt(8), num)
+    ? /^\d/.test(nif)
+      ? NIFType.DNI
+      : NIFType.NIE
+    : NIFType.Unknown;
+};
+
+const getSum = (nif: string) =>
+  nif
+    .substr(1, 7)
+    .split('')
+    .reduce((accum, current, index) => {
+      const num = Number(current) << (index + 1) % 2;
```

Comment:
Only to know it, what is doing the `<<`?

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators